lib/icu.js

```diff
@@ -0,0 +1,209 @@
+        throw new TypeError('Invalid UCS2 Buffer');
+      return icu.convertToUtf8FromUcs2(source);
+    }
+  }
+};
+
+// Transcodes the Buffer from one encoding to another, returning a new
+// Buffer instance.
+exports.transcode = function transcode(source, from_enc, to_enc) {
+  if (!source || !(source.buffer instanceof ArrayBuffer))
+    throw new TypeError('"source" argument must be a Buffer');
+  if (source.length === 0) return Buffer.alloc(0);
+
+  from_enc = normalizeEncoding(from_enc);
+  to_enc = normalizeEncoding(to_enc);
+
+  if (from_enc === to_enc)
+    return Buffer.from(source);
+
+  const cnv_from = conversions[from_enc];
+
+  if (cnv_from) {
+    const cnv_to = cnv_from[to_enc];
+    if (cnv_to)
+      return cnv_to(source);
+  }
+  throw new TypeError(`Unsupported conversion: ${from_enc} to ${to_enc}`);
+};
+
+// Returns the unicode codepoint at a specific offset location in the buffer,
+// based on the given encoding
+exports.codePointAt = function(buf, pos, encoding) {
+  if (!buf || !(buf.buffer instanceof ArrayBuffer))
+    throw new TypeError('First argument must be a Buffer');
+  if (pos < 0 || pos >= buf.length)
+    return;
+  pos >>>= 0;
+
+  encoding = normalizeEncoding(encoding || 'utf8');
+
+  if (encoding === 'binary' || encoding === 'ascii')
+    return buf[pos];
+  if (encoding !== 'utf8' && encoding !== 'ucs2')
+    throw new TypeError(`Unsupported Encoding: ${encoding}`);
+
+  return icu.getCodePointAt(buf, encoding === 'utf8', pos);
+};
+
+// Returns the char at a specific offset location in the buffer, based
+// on the given encoding.
+exports.charAt = function bufferCharAt(buf, pos, encoding) {
+  if (!buf || !(buf.buffer instanceof ArrayBuffer))
+    throw new TypeError('First argument must be a Buffer');
+  if (pos < 0 || pos >= buf.length)
+    return;
+
+  pos >>>= 0;
+  encoding = normalizeEncoding(encoding || 'utf8');
+
+  if (encoding === 'binary' || encoding === 'ascii')
+    return String.fromCharCode(buf[pos]);
+  if (encoding !== 'utf8' && encoding !== 'ucs2')
+    throw new TypeError(`Unsupported Encoding: ${encoding}`);
+
+  return icu.getCharAt(buf, encoding === 'utf8', pos);
+};
+
+function utf8IsLeadByte(c) {
+  return (c & 0xc0) != 0x80;
+}
+
+function utf8CountTrailBytes(c) {
+  return ((c >= 0xc0) + (c >= 0xe0) + (c >= 0xf0));
+}
+
+function utf8ForwardN(buf, start, n) {
+  // Assumes that buf[start] is already aligned on a lead byte
+  const len = buf.length;
+  n >>>= 0;
+  var i = start;
+  if (i >= len) throw new RangeError('index out of range');
+  for (; n > 0; n--)
+    i += 1 + utf8CountTrailBytes(buf[i]);
+  return i;
+}
+
+// Performs a UTF-8 aware slice without error checking. start and end
+// represent character offsets, not byte offsets.
+function utf8Slice(buf, start, end) {
+  if (!buf || !(buf.buffer instanceof ArrayBuffer))
+    throw new TypeError('First argument must be a Buffer');
+  if (typeof end === 'undefined') end = buf.length;
+  start >>= 0;
+  end >>= 0;
```

Comment:
This doesn't operate like `String#slice()` or `Buffer#slice()`. If a negative number is passed then it should slice going backwards. There are some other minor argument coercion differences between this and how Typed Array handles arguments. Buffer should also have its argument parsing reviewed and fixed, but might as well start now. Here's an idea of how they should be parsed:
```js
if (!start) start = 0;
// This works b/c "start < 0" calls valueOf() from start. Evaluate this
// way to be consistent with how Typed Array evaluates. For example:
//   ui8array.subarray(4.6, 5.1): 4.6 -> 4; 5.1 -> 5
//   ui8array.subarray(-3.1, -1.9): -3.1 -> -3; -1.9 -> -1
else start = start < 0 ? Math.ceil(start) : Math.floor(start);

if (end === undefined) end = buf.length;
else if (!end) end = 0;
else end = end < 0 ? Math.ceil(end) : Math.floor(end);

if (end <= start) return Buffer.alloc(0);
```

There's probably be a simpler way to achieve this, but intentionally wanted this to be verbose to easily see what's going on.